File: src/Enterprise/C#/patterns/Strategy.cs
using System;
namespace Genkidama.PatternExamples;
public static class StrategyExample { public static bool Run(){int Price(int v,Func<int,int>s)=>s(v);return Price(100,x=>x)==100&&Price(100,x=>x*80/100)==80;} }

File: src/Enterprise/Java/patterns/strategy.java
import java.util.function.IntUnaryOperator;

class PatternCell {
    private static int price(int base, IntUnaryOperator strategy) {
        return strategy.applyAsInt(base);
    }

    static boolean run() {
        IntUnaryOperator regular = value -> value;
        IntUnaryOperator vip = value -> value * 80 / 100;
        IntUnaryOperator campaign = value -> value >= 100 ? value - 25 : value;

        return price(100, regular) == 100
                && price(100, vip) == 80
                && price(100, campaign) == 75
                && price(80, campaign) == 80;
    }

    public static void main(String[] args) {
        if (!run()) {
            throw new AssertionError("Java Strategy contract failed");
        }
    }
}
